### PR TITLE
Crank up invalid value lint

### PR DESCRIPTION
**Pull request by RalfJung**
* Warn against uninit `bool` and `char`.
* Warn against 0-init `NonNull` and friends
* Detect transmute-from-0 as zero-initialization ([seen in the wild](https://github.com/glium/glium/issues/1775#issuecomment-522144636))

### CODE REVIEW

File: src/librustc_lint/builtin.rs [hunk 1]
```diff
@@ -1876,25 +1876,103 @@ declare_lint_pass!(InvalidValue => [INVALID_VALUE]);
...
+            const ZEROED_PATH: &[Symbol] = &[sym::core, sym::mem, sym::zeroed];
+            const UININIT_PATH: &[Symbol] = &[sym::core, sym::mem, sym::uninitialized];
+            // `transmute` is inside an anonymous module (the `extern` block?);
+            // `Invalid` represents the empty string and matches that.
+            const TRANSMUTE_PATH: &[Symbol] =
+                &[sym::core, sym::intrinsics, kw::Invalid, sym::transmute];
+
+            if let hir::ExprKind::Call(ref path_expr, ref args) = expr.node {
+                if let hir::ExprKind::Path(ref qpath) = path_expr.node {
+                    if let Some(def_id) = cx.tables.qpath_res(qpath, path_expr.hir_id)
```
Comment (Centril) [inline src/librustc_lint/builtin.rs]:
```suggestion
                    let def_id = cx.tables.qpath_res(qpath, path_expr.hir_id).opt_def_id()?;
```
A little less rightward drift. =)

File: src/librustc_lint/builtin.rs [hunk 2]
```diff
@@ -1876,25 +1876,103 @@ declare_lint_pass!(InvalidValue => [INVALID_VALUE]);
...
+        ) -> Option<InitError> {
             use rustc::ty::TyKind::*;
             match ty.sty {
                 // Primitive types that don't like 0 as a value.
                 Ref(..) => Some((format!("References must be non-null"), None)),
                 Adt(..) if ty.is_box() => Some((format!("`Box` must be non-null"), None)),
                 FnPtr(..) => Some((format!("Function pointers must be non-null"), None)),
                 Never => Some((format!("The never type (`!`) has no valid value"), None)),
-                // Recurse for some compound types.
+                // Primitive types with other constraints
```
Comment (Centril) [inline src/librustc_lint/builtin.rs]:
```suggestion
                // Primitive types with other constraints.
```

File: src/librustc_lint/builtin.rs [hunk 3]
```diff
@@ -1876,25 +1876,103 @@ declare_lint_pass!(InvalidValue => [INVALID_VALUE]);
...
+                Bool if init == InitKind::Uninit =>
+                    Some((format!("Booleans must be `true` or `false`"), None)),
+                Char if init == InitKind::Uninit =>
+                    Some((format!("Characters must be a valid unicode codepoint"), None)),
+                // Recurse and checks for some compound types.
                 Adt(adt_def, substs) if !adt_def.is_union() => {
```
Comment (Centril) [inline src/librustc_lint/builtin.rs]: Consider moving parts (the two large sections) or the whole arm into its own function; it's getting a bit long.

**Comment by Centril** [comment]
r? @Centril 

r=me with comments addressed :)

File: src/librustc_lint/builtin.rs (hunk 1 above)
Comment (RalfJung) [reply src/librustc_lint/builtin.rs]: Nice catch!

File: src/librustc_lint/builtin.rs (hunk 3 above)
Comment (RalfJung) [reply src/librustc_lint/builtin.rs]: TBH I feel this is still reasonable. The individual parts as their own function don't feel very self-contained.

Comment (Centril) [reply src/librustc_lint/builtin.rs]: Alright :)

File: src/librustc_lint/builtin.rs [hunk 4]
```diff
@@ -1876,25 +1876,101 @@ declare_lint_pass!(InvalidValue => [INVALID_VALUE]);
...
+            // `transmute` is inside an anonymous module (the `extern` block?);
+            // `Invalid` represents the empty string and matches that.
+            const TRANSMUTE_PATH: &[Symbol] =
+                &[sym::core, sym::intrinsics, kw::Invalid, sym::transmute];
+
+            if let hir::ExprKind::Call(ref path_expr, ref args) = expr.node {
+                if let hir::ExprKind::Path(ref qpath) = path_expr.node {
+                    let def_id = cx.tables.qpath_res(qpath, path_expr.hir_id).opt_def_id()?;
+
+                    if cx.match_def_path(def_id, &ZEROED_PATH) {
```
Comment (lzutao) [inline src/librustc_lint/builtin.rs]: No need to borrow ZEROED_PATH here.

Comment (RalfJung) [reply src/librustc_lint/builtin.rs]: Good point, thanks.

**Comment by RalfJung** [comment]
Ah, this is caused by musl not being able to display spans in libstd... I cannot test with `Vec`. :/

**Pull request merged**